### ISSUE TITLE
Block-Diagonal Covariance

### ISSUE THREAD

**Issue by JonathanWenger** [inferno-ml/inferno#10]
Block-diagonal covariances allow independent parameters that are all part of a single ``torch.nn.Parameter``, for example in a ``bnn.Linear`` layer. This is a covariance structure that comes up quite a bit and is useful for efficient variational networks.

Common examples are to ensure that attention heads in ``bnn.MultiheadAttention`` are independent or that the logits in classification are independent across classes.
